Nginx : Get rid of "ff_host" thread

firesidefox opened the pull request:
Get rid of "ff_host" thread dealing with the kernel (including channel). So there will be one thread for each nginx worker. And use directive "schedule_timeout" to adjust poll frequency of kernel.
This is a better way to ```support kernel network stack #117```

Commit:
Nginx : Get rid of "ff_host" thread, so single thread is used for both fstack and kernel.

whl739 commented on app/nginx-1.11.10/src/event/ngx_event.c:
```diff
@@ -239,7 +251,19 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
     delta = ngx_current_msec;
 
-    (void) ngx_process_events(cycle, timer, flags);
+    //(void) ngx_process_events(cycle, timer, flags);
```
Why `ngx_process_events` is commented?

firesidefox replied:
I comment it for comparison test, sorry that you see it. 
Should I close this PR, and create a new one?

whl739 commented:
Does this patch have any effect on performance?

firesidefox commented:
1. Performance is almost the same as previous. 
2. It has made least modification to the nginx source code.
3. It has contributed to a mixed proxy between fstack and kernel.
```
server {
    kernel_network_stack off;
    listen       80;
    location / {
        root html;
        index  index.html index.htm;
    }
    location /proxy {
        # we can connect to localhost with kernel stack by directive below.
        proxy_kernel_network_stack on;
        proxy_pass http://127.0.0.1:8000;
    }
}
```

I will  quickly create a PR to add directive ```proxy_kernel_network_stack```.

firesidefox committed:
Nginx: Restore unexpected Comment

whl739 commented:
Ok. I'll review the code and test next week.

whl739 commented on app/nginx-1.11.10/src/event/ngx_event.c:
```diff
@@ -241,6 +253,18 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
+        initial = ngx_current_msec;
+        tick = 0;
+    }
+    tick++;
```
There seems to be a problem here, `ngx_schedule_timeout`'s unit is millisecond, but `tick` is not, every main loop iteration, `tick` will plus 1, the comparison `tick >= ngx_schedule_timeout` is unreasonable.

firesidefox replied:
Nginx reduces ```gettimeofday``` system calls by caching time variables to improve efficiency. And ```ngx_current_msec``` would not be updated in each loop iteration. So ```tick``` is designed to avoid inability to poll host ,when there is no event.

I just assume that a loop iteration is equal to one millisecond. Do you mean this is a unreasonable conversion?
I want to modify the comparison to ```tick >= ngx_schedule_timeout * n``` (for one ```n =1000```), Do you agree with this?

whl739 replied:
> I just assume that a loop iteration is equal to one millisecond. Do you mean this is a unreasonable conversion?
I want to modify the comparison to tick >= ngx_schedule_timeout * n (for one n =1000), Do you agree with this?

F-Stack is running on polling mode, the time of one loop iteration is much less than one millisecond, its value is related to the performance of CPU and the number of events in this iteration. It's not fixed.

> Nginx reduces gettimeofday system calls by caching time variables to improve efficiency. And ngx_current_msec would not be updated in each loop iteration. So tick is designed to avoid inability to poll host ,when there is no event.

I consider that `ngx_current_msec` is ok, nginx's related code is below:
```
void
ngx_process_events_and_timers(ngx_cycle_t *cycle)
{
    ngx_uint_t  flags;
    ngx_msec_t  timer, delta;

    if (ngx_timer_resolution) {
        timer = NGX_TIMER_INFINITE;
        flags = 0;

    } else {
        timer = ngx_event_find_timer();
        flags = NGX_UPDATE_TIME;
        ....
   }
   .....
}
```
if `NGX_UPDATE_TIME` is set, `ngx_current_msec` will be updated.

firesidefox replied:
I reanalyze the code, and see your point about this.

whl739 commented:
Note that `doc/F-Stack_Nginx_APP_Guide.md` should be updated.

firesidefox committed:
Nginx: Cut out code about poll-kernel frequency

Cut out unreasonable code about poll-kernel frequency

whl739 commented:
Thanks for your contribution. I'll test it next week.
And about the `doc/F-Stack_Nginx_APP_Guide.md`, it's better to add some descriptions of `kernel_network_stack `.

firesidefox commented:
I've already added now.

whl739 commented:
When nginx is configured like this:
```
server {
    listen 80;
    kernel_network_stack on;
    location / {
        proxy_pass http://127.0.0.1:8080/;
    }
}
```
this will cause an error: `connect() to 127.0.0.1:8080 failed (99: Cannot assign requested address)`, because now the host events and ff events are all in one thread, sockets are all created by `ff_socket`, but in the previous version, the host event thread's sockets are created by the syscall `socket`.

whl739 commented:
And another question, i think the changes in `ngx_get_connection` and `ngx_free_connection` are no longer necessary, what's your opinion?

firesidefox commented:
A directive ```proxy_kernel_network_stack``` will be added very soon and it determines which network stack is used by ```proxy``` to set up connections.

About ```ngx_get_connection``` and ```ngx_free_connection```, I agree with you.

whl739 commented:
Will `proxy_kernel_network_stack` be pushed to this pr or another?

Commit:
Nginx: Restore ngx_get_connection and ngx_free_connection

The code that we used to preserve ngx_get_connection and ngx_free_connection thread safety , is not necessary now.

firesidefox commented:
Another, because I think the function of this pr is complete and available.
